airbyte-lib/airbyte_lib/datasets/_base.py

```diff
@@ -11,11 +11,24 @@
 class DatasetBase(ABC):
     """Base implementation for all datasets."""
 
+    def __init__(self) -> None:
+        self._length: int | None = None
+
     @abstractmethod
     def __iter__(self) -> Iterator[Mapping[str, Any]]:
         """Return the iterator of records."""
         raise NotImplementedError
 
+    def __len__(self) -> int:
```

Comment:
This breaks the lazy dataset. When calling `len(lazy_dataset)`, it will work correctly, but `list(lazy_dataset)` will not give any results anymore, I think it's because the sum(1 for _ in self) is exhausting the iterator which is of course bad if you want to actually read the data (the implementation in lazy dataset is not creating a new iterator when calling `__iter__`)

I'm not sure about the "right" solution, but we need to make sure that getting length will not affect the state of the dataset instance.

Reply:
Good catch. I am trying `self._length = sum(1 for _ in iter(self))` as a next iteration which _should_ in theory generate a new iterator from self and not exhaust the (self) itertor.

Reply:
@flash1293 - I didn't get the iterator issue fixed and so for expediency, I've just moved it into SQL-based datasets only. This is also where we can tune it by running count(*) on the database, so I think it makes sense now.

Then, for lazy datasets, users will still need to use the len(list(...)) approach for now.